Sprint 3/step 1

OlegTar opened the pull request.

OlegTar committed:
модалка

OlegTar committed:
страницы логин, восстановление пароль, сброс пароля, регистрация

OlegTar committed:
страница профиля

OlegTar committed:
логин, логаут, регистрация, редактирование пользователя.

OlegTar committed:
исправлена опечатка

OlegTar committed:
рефакторинг

OlegTar committed:
рефакторинг

OlegTar committed:
последние штрихи

OlegTar committed:
Чтобы ревьювер вдруг не подумал, что у меня не в proteted-route.tsx проверяется пользаки

gennady-bars commented on src/services/actions/password-reset.ts:
```diff
@@ -0,0 +1,24 @@
+			await sendPasswordResetRequest(passwordResetRequest);
+			dispatch(setSuccess());
+			localStorage.setItem(resetPasswordCalled, '1');
+		} catch (e) {
```
В `createAsyncThunk` не нужно указывать блок `catch`, так как `createAsyncThunk` сам ловит ошибки и отображает их в поле `rejected`. А его уже обрабатывать можно в `createSlice`

OlegTar replied:
Хорошо, я переделаю, но наставник нам дал два примера, в одном используется action.rejected, в другом делается .then .catch в createAsyncThunk (что аналог try/catch, если использовать async/await)

gennady-bars commented on src/services/actions/password-reset.ts:
```diff
@@ -0,0 +1,24 @@
+	'user/passwordReset',
+	async (passwordResetRequest: PasswordResetRequest, { dispatch }) => {
+		try {
+			dispatch(setPending());
```
В `createAsyncThunk` не нужно указывать `setPending`, так как `createAsyncThunk` сам отслеживает загрузку через `pending` в слайсе. А его уже обрабатывать можно в `createSlice`

gennady-bars commented on src/pages/forgot-password/forgot-password.tsx:
```diff
@@ -0,0 +1,89 @@
+		state: state.user.state,
+		errorMessage: state.user.errorMessage,
+	}));
+	const [email, setEmail] = useState('');
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

OlegTar replied:
Спасибо, заюзал

gennady-bars commented on src/pages/forgot-password/forgot-password.tsx:
```diff
@@ -0,0 +1,89 @@
+		<>
+			<RequestStatus state={state} errorMessage={errorMessage} />
+			<form
+				onSubmit={(e) => {
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,51 @@
+		);
+	}
+
+	if (onlyUnAuth && isAuthenticated) {
+		// Пользователь авторизован, но роут предназначен для неавторизованного пользователя
+		// Делаем редирект на главную страницу или на тот адрес, что записан в location.state.from
+		const { from } = location.state || { from: { pathname: '/' } };
+		return <Navigate to={from} />;
+	}
+
+	if (!onlyUnAuth && !isAuthenticated) {
+		return <Navigate to="/login" state={{ from: location }} />;
```
Хорошо сделан `Protected`

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,19 +1,34 @@
+	const location = useLocation();
+	const navigate = useNavigate();
+
+	const background = location.state && location.state.background;
```
Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -33,16 +57,72 @@ function App() {
-					<BurgerIngredients />
-					<BurgerConstructor />
-				</DndProvider>
+				<Routes>
```
Вот тут нужно добавить `location`, чтобы под попапом отображалась страница

OlegTar replied:
можно пояснить?

OlegTar replied:
@gennady-bars можно пояснить, что за location ?

OlegTar replied:
location={background || location}
вопрос решен. у меня правда работало и без него, потому что я в navigate state проставлял

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:


* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. 
* Хорошо сделан `Protected`
* Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно

но есть некоторые недочеты:


* Сквозь **оверлей** должна проглядываться страница, так как попап открывается **над** ней. Она не должна никуда пропадать, иначе это приводит к перерисовке всего контента
* В `createAsyncThunk` не нужно указывать блок `catch`, так как `createAsyncThunk` сам ловит ошибки и отображает их в поле `rejected`. А его уже обрабатывать можно в `createSlice`
* В `createAsyncThunk` не нужно указывать `setPending`, так как `createAsyncThunk` сам отслеживает загрузку через `pending` в слайсе. А его уже обрабатывать можно в `createSlice`
* 


**Можно лучше**


* при клике на **логотип** обычно идет переход на **главную** страницу

OlegTar committed:
замечания по пуллреквесту (пока не все)

OlegTar committed:
Переход на главную страницу при лого

OlegTar committed:
добавлен location

OlegTar committed:
выносим работу с localStorage в api, чтобы единообразно

OlegTar committed:
удален лишний код. он не отрабатывал, так как protected-route делал своё дело.

OlegTar committed:
на всякий случай верну сделаю Navigate в login.tsx, хотя и без него работало, в protected-route

OlegTar committed:
убран лишний код.

модалка "загрузка" в protected-route не нужна, у меня и так на каждый реквест показывается "Загрузка"
Во-вторых, пусть показывается загрузка со странице "login'а сзади"
удалил свойство authChecked
удалил экшен setAuthChecked, все действия в редьюсере.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
